tests/mir-opt/box_conditional_drop_allocator.rs

```diff
@@ -0,0 +1,38 @@
+// skip-filecheck
+//@ test-mir-pass: ElaborateDrops
+#![feature(allocator_api)]
+
+// Regression test for #131082.
+// Testing that the allocator of a Box is dropped in conditional drops
+
+use std::alloc::{AllocError, Allocator, Global, Layout};
+use std::ptr::NonNull;
+
+struct DropAllocator;
+
+unsafe impl Allocator for DropAllocator {
+    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        Global.allocate(layout)
+    }
+    unsafe fn deallocate(&self, ptr: NonNull<u8>, layout: Layout) {
+        Global.deallocate(ptr, layout);
+    }
+}
+impl Drop for DropAllocator {
+    fn drop(&mut self) {}
```

Comment:
It might be good to have this function call something that leaves a trace in the MIR during `ElaborateDrops` since right now, the no-op drop just results in an uninteresting chain of gotos. 

To avoid polluting the MIR too much, you can probably do something like this:

```rust
fn drop(&mut self) {
    extern { fn drop_allocator_drop(); }
    unsafe { drop_allocator_drop(); }
}
```

which would then just show up as a `_tmp = drop_allocator_drop() -> [...]` without much extra stuff around it.

Reply:
I tried this, but it didn't actually change the MIR. I don't think the drops are being inlined, the gotos are just an artifact of how drop elaborate injects its drop ladders into the existing MIR.